Buy and Sell Stock once

In [4]:
arr = [100, 180, 260, 310, 40, 535, 695]
max_profit = 0
for i in range(len(arr)):
    for j in range(i+1, len(arr), 1):
        if arr[j] - arr[i] > max_profit:
            max_profit = arr[j] - arr[i]
            print("max_profit: {} for element: {}".format(max_profit, arr[i]))

print(max_profit)
# time complexity - O(n^2)
# space complexity - O(1)


max_profit: 80 for element: 100
max_profit: 160 for element: 100
max_profit: 210 for element: 100
max_profit: 435 for element: 100
max_profit: 595 for element: 100
max_profit: 655 for element: 40
655


The cost of a stock on each day is given in an array. 
Find the maximum profit that you can make by buying and selling on those days. 
If the given array of prices is sorted in decreasing order, then profit cannot be earned at all.

link - https://www.geeksforgeeks.org/stock-buy-sell/ 

arr = [100, 180, 260, 310, 40, 535, 695]
ouput - 865
Explanation - Buy the stock on day 0 and sell it on day 3 => 310 – 100 = 210
            Buy the stock on day 4 and sell it on day 6 => 695 – 40 = 655
            Maximum Profit  = 210 + 655 = 865

In [11]:
maxProfit = 0
minPrice = float('inf') 

for i in range(len(arr)):
    minPrice = min(minPrice, arr[i])
    maxProfit = max(maxProfit, arr[i]-minPrice)
        
print(maxProfit)

# time complexity - O(n)
# space complexity - O(1)
    

655
